### ISSUE TITLE
Interface injection fails to resolve constructor argument

### ISSUE THREAD

**Issue by dingo-d** [PHP-DI/PHP-DI#658]
I'm trying to set up constructor injection using interface typehints but it keeps on failing.

I have an `Aws_Client` class that implements `Client` interface. This is a wrapper for my `aws-sdk-php`.

My `Admin_Credentials` class, extends the `Base_Ajax_Callback` abstract class that implements two interfaces - `Service` and `Invokable` interfaces. `Service` interface has a `register` method (used for injecting WordPress hooks), while `Invokable` interface has `callback` method that is used for defining ajax callback.

The `Admin_Credentials` class needs to use some functionality from `Aws_Client` class, so I added this in a constructor

```php
<?php

namespace My_Plugin\Callback;

use My_Plugin\Api\Client:

class Admin_Credentials extends Base_Ajax_Callback {
  
  protected $aws_client;

  public function __construct( Client $aws_client ) {
    $this->aws_client = $aws_client;
  }

  public function callback() {
    // ... some business logic goes here
  }

}
```

In my plugin entrypoint I have

```php
$builder = new ContainerBuilder();
$builder->useAnnotations( false );
$builder->useAutowiring( true );

$builder->addDefinitions(
  [
    Callback\Invokable::class => \DI\create( Callback\Admin_Credentials::class )
      ->constructor( Api\Aws_Client_Service::class ),
  ]
);

$container = $builder->build();
```

This will throw

```bash
[21-May-2019 10:39:28 UTC] PHP Fatal error:  Uncaught DI\Definition\Exception\InvalidDefinition: Entry "My_Plugin\Callback\Admin_Credentials" cannot be resolved: Entry "My_Plugin\Api\Client" cannot be resolved: the class is not instantiable
Full definition:
Object (
    class = #NOT INSTANTIABLE# My_Plugin\Api\Client
    lazy = false
)
Full definition:
Object (
    class = My_Plugin\Callback\Admin_Credentials
    lazy = false
    __construct(
        $aws_client = get(My_Plugin\Api\Client)
    )
) in .../vendor/php-di/php-di/src/Definition/Exception/InvalidDefinition.php:18
Stack trace:
#0 .../vendor/php-di/php-di/src/Definition/Resolver/ObjectCreator.php(155): DI\Definition\Exception\InvalidDefinition::create(Object(DI\Definition\ObjectDefinition), 'Entry "My_Plugin...')
#1 .../vendor/php-di/php-di/src/Defi in .../vendor/php-di/php-di/src/Definition/Exception/InvalidDefinition.php on line 18
```

The same error is created if I don't add definitions for my builder.

I've also tried with 

```php
$builder = new ContainerBuilder();
$builder->useAnnotations( false );
$builder->useAutowiring( true );

$builder->addDefinitions(
  [
    Service::class => \DI\autowire( Callback\Admin_Credentials::class )
      ->constructorParameter( Api\Client::class, Api\Aws_Client_Service::class ),
  ]
);

$container = $builder->build();
```

No matter what I try, it looks like the container expects a concrete class (the ` $aws_client = get(My_Plugin\Api\Client)` part). 

I looked at this [question on SO](https://stackoverflow.com/questions/51957708/php-di-interface-being-injected-into-constructor-wont-resolve-correctly) and I should be able to do this (even without specifying the constructor parameters).

I don't want to do this:

```php
<?php

namespace My_Plugin\Callback;

use My_Plugin\Api\Aws_Client:

class Admin_Credentials extends Base_Ajax_Callback {
  
  protected $aws_client;

  public function __construct( Aws_Client $aws_client ) {
    $this->aws_client = $aws_client;
  }
```

As this couples the `Admin_Credentials` class with the concrete implementation of the `Client` interface - and that makes it hard for me to mock during unit and integration tests.

**Comment by mnapoli** [comment]
```
Uncaught DI\Definition\Exception\InvalidDefinition:
  Entry "My_Plugin\Callback\Admin_Credentials" cannot be resolved:
    Entry "My_Plugin\Api\Client" cannot be resolved: the class is not instantiable
```

It seems your code tries to `$container->get()` the `My_Plugin\Callback\Admin_Credentials` entry, **not** `Callback\Invokable` (which isn't mentioned in the error).

You have configured `Callback\Invokable`, but `My_Plugin\Callback\Admin_Credentials` is not configured. I think this is the root of the error.

**Comment by dingo-d** [comment]
So I need to specify in my container the concrete class I'm passing in the constructor in my definitions like

```php
    $builder->addDefinitions(
      [
        Callback\Admin_Credentials::class => \DI\autowire( Api\Aws_Client_Service::class ),
      ]
    );
```

?

**Comment by mnapoli** [comment]
Not necessarily.

You need to specify in the container what is *type-hinted* in your constructor.

**Comment by dingo-d** [comment]
I tried that (I'm looking at the documentation and working by those examples) but that's not working :/

Or I'm doing something wrong. Do you have an example how this should look like?

**Comment by SvenRtbg** [comment]
First of all it is making it very hard to follow your code because you do not use fully qualified class names, and you do omit any `use` clauses from the top of the files. This makes it impossible to see if `Callback\Admin_Credentials` references the class you were talking about earlier, or if it points to a different class.

My general recommendation would be to always use fully qualified class names in the DI configuration because plenty of class names will appear there - the chance of two class names with the same suffix (like `Client` to get there is high, so you'd have to use alias names when importing, further adding to confusion, and you'd be importing about all classes anyways, which will double the lines in the `use` block.

**Comment by dingo-d** [comment]
I'm aliasing the 'parent' classes at the top of my plugin entry point, and I have LSP set up in my sublime, so I can check if the classes are loaded properly.

I am using `classmap` autoloading in composer, that shouldn't affect the autowiring, right?

**Comment by mnapoli** [comment]
Here is an example on how to use it, does it help?

```php
$builder = new ContainerBuilder();
$builder->useAnnotations( false );
$builder->useAutowiring( true );

$builder->addDefinitions(
  [
    Callback\Invokable::class => \DI\create( Callback\Admin_Credentials::class )
      ->constructor( Api\Aws_Client_Service::class ),
  ]
);

$container = $builder->build();

class UserController
{
    public function __construct(\Callback\Invokable $parameter)
    { /* ...*/ }
}
```

**Comment by dingo-d** [comment]
I tried that in the original case (original comment). It's producing the same error

```bash

[21-May-2019 13:02:29 UTC] PHP Fatal error:  Uncaught DI\Definition\Exception\InvalidDefinition: Entry "My_Plugin\Callback\Admin_Credentials" cannot be resolved: Entry "My_Plugin\Api\Client" cannot be resolved: the class is not instantiable
Full definition:
Object (
    class = #NOT INSTANTIABLE# My_Plugin\Api\Client
    lazy = false
)
Full definition:
Object (
    class = My_Plugin\Callback\Admin_Credentials
    lazy = false
    __construct(
        $aws_client = get(My_Plugin\Api\Client)
    )
) in /vendor/php-di/php-di/src/Definition/Exception/InvalidDefinition.php:18
Stack trace:
#0 /vendor/php-di/php-di/src/Definition/Resolver/ObjectCreator.php(155): DI\Definition\Exception\InvalidDefinition::create(Object(DI\Definition\ObjectDefinition), 'Entry "My_Plugin...')
#1 /vendor/php-di/php-di/src/Defi in /vendor/php-di/php-di/src/Definition/Exception/InvalidDefinition.php on line 18
```

It works with concrete implementation in the constructor, not interface 🤷‍♂

**Comment by mnapoli** [comment]
OK could you show how the `` is fetched from the container? (how is it created?)

Also as a side note: `->constructor( Api\Aws_Client_Service::class )` this will not work. You need to use `get(...class name...)` if you want to reference another service. Here you are simply passing a string (the class name) to the constructor. But this bug isn't visible yet because of the root bug: you don't even get to that yet. Still it's best to fix it now.

**Comment by dingo-d** [comment]
Ok, so basically my plugin class where I instantiate all my classes look like this:

```php
<?php
use \DI\ContainerBuilder;
...

final class Plugin implements Registerable, Has_Activation, Has_Deactivation {

    private $services = [];

    public function register_services() {
        // Bail early so we don't instantiate services twice.
        if ( ! empty( $this->services ) ) {
            return;
        }

        $builder = new ContainerBuilder();

        $container = $builder->build();

        $this->services = array_map(
            function( $class ) use ( $container ) {
                return $container->get( $class );
            },
            $this->get_service_classes()
        );
       
        // register hooks
        array_walk(
            $this->services,
            function( $class ) {
                if ( ! $class instanceof Service ) {
                    throw Exception\Invalid_Service::from_service( $class );
                }

                $class->register();
            }
        );
    }
}
```

`get_service_classes` method is just a list of FQCNs

```php
private function get_service_classes() : array {
    return [
      Admin\Optimizations::class,
      Users\User_Manager::class,
            ...
    ];
}```

When I have services without constructor injection, this works perfectly. When I need to add DI...

**Comment by dingo-d** [comment]
Ok I figured it out. Instead of doing this:

```php
$builder->addDefinitions(
  [
    Callback\Invokable::class => \DI\create( Callback\Admin_Credentials::class )
      ->constructor( Api\Aws_Client_Service::class ),
  ]
);
```

Which is essentially passing a **string** to a constructor, I needed to do this

```php
$builder->addDefinitions(
  [
    Callback\Admin_Credentials::class => \DI\create()
      ->constructor( \DI\get( Api\Aws_Client_Service::class ) ),
  ]
);
```

Where I'm actually passing a class to the constructor. this works 🎉 

Thanks for the help!

**Comment by lezhnev74** [comment]
Stumbled upon the same problem. 
Maybe we can add this detail to the docs?

I was trying to configure lazy loaded classes like this:
```php
\Textsite\Infrastructure\HTTP\ErrorHandler::class => \DI\create()->lazy(),
```
That did not work, as `ErrorHandler` needs some dependencies in the constructor.

Then I tried:
```php
\Textsite\Infrastructure\HTTP\ErrorHandler::class => \DI\create()
        ->constructor(
            \Psr\Log\LoggerInterface::class,
            \eftec\bladeone\BladeOne::class
        )
        ->lazy(),
```
No luck.

Finally, this worked:
```php
\Textsite\Infrastructure\HTTP\ErrorHandler::class => \DI\create()
        ->constructor(
            \DI\get(\Psr\Log\LoggerInterface::class),
            \DI\get(\eftec\bladeone\BladeOne::class)
        )
        ->lazy(),
```

**Comment by mnapoli** [comment]
@lezhnev74 use `autowire()` instead of `create()` to benefit from autowiring.

**Comment by lezhnev74** [comment]
Thanks @mnapoli, will do! It wasn't clear for me from the docs that `autowire()` also supports lazy loading.
